pkg/sql/colexec/execgen/template.go

```diff
@@ -0,0 +1,351 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package execgen
+
+import (
+	"fmt"
+	"go/token"
+	"regexp"
+	"strings"
+
+	"github.com/dave/dst"
+	"github.com/dave/dst/dstutil"
+)
+
+type templateParamInfo struct {
+	fieldOrdinal int
+	field        *dst.Field
+	instances    []*dst.Expr
+}
+
+type funcInfo struct {
+	decl           *dst.FuncDecl
+	templateParams []templateParamInfo
+}
+
+// Match // execgen:template<foo, bar>
+var templateRe = regexp.MustCompile(`\/\/ execgen:template<((?:(?:\w+),?\W*)+)>`)
+
+// replaceTemplateVars removes the template arguments from a callsite of a
+// templated function. It returns the template arguments that were used, and a
+// new CallExpr that doesn't have the template arguments.
+func replaceTemplateVars(
+	info *funcInfo, call *dst.CallExpr,
+) (templateArgs []dst.Expr, newCall *dst.CallExpr) {
+	if len(info.templateParams) == 0 {
+		return nil, call
+	}
+	templateArgs = make([]dst.Expr, len(info.templateParams))
+	// Collect template arguments.
+	for i, param := range info.templateParams {
+		templateArgs[i] = dst.Clone(call.Args[param.fieldOrdinal]).(dst.Expr)
+	}
+	// Remove template vars from callsite.
+	newArgs := make([]dst.Expr, 0, len(call.Args)-len(info.templateParams))
```

Comment:
That's wierd -- I got panics when I copied the args from the original expression instead of cloning it first, and then removing the other args.